python/ray/data/_internal/execution/operators/map_data_processor.py

```diff
@@ -0,0 +1,319 @@
+import functools
+import itertools
+from enum import Enum
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+
+from ray.data._internal.block_batching.block_batching import batch_blocks
+from ray.data._internal.execution.interfaces.task_context import TaskContext
+from ray.data._internal.output_buffer import BlockOutputBuffer
+from ray.data.block import Block, BlockAccessor, DataBatch
+from ray.data.context import DataContext
+
+Row = Dict[str, Any]
+MapTransformFnData = Union[Block, Row, DataBatch]
+MapTransformCallable = Callable[
+    [Iterable[MapTransformFnData], TaskContext], Iterable[MapTransformFnData]
+]
+BlockTransformCallable = Callable[[Iterable[Block], TaskContext], Iterable[Block]]
+RowTransformCallable = Callable[[Iterable[Row], TaskContext], Iterable[Row]]
+BatchTransformCallable = Callable[
+    [Iterable[DataBatch], TaskContext], Iterable[DataBatch]
+]
+InitFn = Callable[[], None]
+
+
+class MapTransformFnDataType(Enum):
+    """An enum that represents the input/output data type of a MapTransformFn."""
+
+    Block = 0
+    Row = 1
+    Batch = 2
+
+
+class MapTransformFn:
+    """Represents a single transform function in a MapDataProcessor."""
+
+    def __init__(
+        self,
+        callable: MapTransformCallable,
+        input_type: MapTransformFnDataType,
+        output_type: MapTransformFnDataType,
+    ):
+        """
+        Args:
+            callable: the underlying Python callable object.
+            input_type: the type of the input data.
+            output_type: the type of the output data.
+        """
+        self._callable = callable
+        self._input_type = input_type
+        self._output_type = output_type
+
+    def __call__(
+        self, input: Iterable[MapTransformFnData], ctx: TaskContext
+    ) -> Iterable[MapTransformFnData]:
+        return self._callable(input, ctx)
+
+    @property
+    def input_type(self) -> MapTransformFnDataType:
+        return self._input_type
+
+    @property
+    def output_type(self) -> MapTransformFnDataType:
+        return self._output_type
+
+
+class MapDataProcessor:
```

Comment:
I agree `MapDataProcessor` is a bit too general. But I'd prefer not to use terms like "composer" or "chain", because they expose unnecessary implementation details. From the callers' perspective, what matters is that this class processes (or transforms) data. 
If you want to avoid confusion with preprocessors? what about "MapDataTransformer".

Reply:
sounds good